### PR TITLE
Fix activation checkpoint metadata loss through custom autograd.Function

### PR DESCRIPTION
**Pull request by soulitzer**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #175338
* #176447
* #176455
* #175348
* #176931
* __->__ #177396
* #174327

Problem:

When a checkpoint region contains a custom autograd.Function, the ops
inside forward() were not getting ac_graph_id or recompute tags in the
joint graph. This happened because tag_nodes only sees the top-level
graph nodes, but autograd_function_apply is an opaque HOP whose
subgraph nodes are never visited. When the HOP is later desugared
during AOTAutograd tracing, the inlined ops have no checkpoint tags,
so the partitioner saves their intermediates instead of recomputing
them — defeating activation checkpointing entirely.

Solution:

PR fixes it by unifying the plain-AC case (which previously relied on tagging before autograd.Function HOPs are desugared) to now also through SAC path which annotates using TorchDispatchMode. This PR also makes sure unified path applies ac_graph_id properly (previously ac_graph_id tagging relied on tagging)

Authored with Claude.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @kadeng @chauhang @amjames @Lucaskabela @jataylo @mlazos

### CODE REVIEW

File: torch/_higher_order_ops/wrap.py [hunk 1]
```diff
@@ -434,44 +435,122 @@ def __call__(self, gmod, *args, **kwargs):
 tag_activation_checkpoint = TagActivationCheckpoint()
 
 
+class _ACTaggingTorchDispatchMode(torch.utils._python_dispatch.TorchDispatchMode):
```
Comment (Skylion007) [inline torch/_higher_order_ops/wrap.py]:
```suggestion
class _ACTaggingTorchDispatchMode(torch.utils._python_dispatch.TorchDispatchMode, Generic[_P, _R]):

File: torch/_higher_order_ops/wrap.py [hunk 2]
```diff
@@ -434,44 +435,122 @@ def __call__(self, gmod, *args, **kwargs):
...
+
+    @classmethod
+    def ignore_compile_internals(cls):
+        return True
+
+    def __init__(self, ac_graph_id, policy_fn=None):
+        self.ac_graph_id = ac_graph_id
+        self.policy_fn = policy_fn
+
+    def __torch_dispatch__(self, func, types, args=(), kwargs=None):
```
Comment (Skylion007) [inline torch/_higher_order_ops/wrap.py]:
```suggestion
    def __torch_dispatch__(self, func : _P2, types, args: _P2.args =(), kwargs : _P2.kwargs | Optional =None):

File: torch/_higher_order_ops/wrap.py [hunk 3]
```diff
@@ -434,44 +435,122 @@ def __call__(self, gmod, *args, **kwargs):
...
+    For SAC, policy_fn is the user's selective checkpoint policy.
+    """
+
+    supports_higher_order_operators = True
+
+    @classmethod
+    def ignore_compile_internals(cls):
+        return True
+
+    def __init__(self, ac_graph_id, policy_fn=None):
```
Comment (Skylion007) [inline torch/_higher_order_ops/wrap.py]:
```suggestion
    def __init__(self, ac_graph_id, policy_fn : Callable[_P, _R]=None):
```

File: torch/_higher_order_ops/wrap.py (hunk 1 above)
Comment (Skylion007) [reply torch/_higher_order_ops/wrap.py]: You can have some great parameter typing here to make sure the policy Function specified through the dispatch matches the API of the function dispatched through __torch__function__

File: test/dynamo/test_activation_checkpointing.py [hunk 4]
```diff
@@ -2240,16 +2288,16 @@ def forward(self, arg0_1, arg1_1):
...
     t = torch.ops.aten.t.default(arg0_1);  arg0_1 = None
     mm_2 = torch.ops.aten.mm.default(t, sigmoid_backward);  t = None
     t_1 = torch.ops.aten.t.default(arg1_1);  arg1_1 = None
     mm_3 = torch.ops.aten.mm.default(sigmoid_backward, t_1);  sigmoid_backward = t_1 = None
-    detach_3 = torch.ops.aten.detach.default(mm_3);  mm_3 = None
-    detach_4 = torch.ops.aten.detach.default(mm_2);  mm_2 = None
-    return (detach_3, detach_4)""",
+    detach_5 = torch.ops.aten.detach.default(mm_3);  mm_3 = None
+    detach_6 = torch.ops.aten.detach.default(mm_2);  mm_2 = None
+    return (detach_5, detach_6)""",
```
Comment (soulitzer) [inline test/dynamo/test_activation_checkpointing.py:2358]: The detach numbering changes in tests like test_ac_rematerialize_simple_forward_backward are because vanilla AC now goes through _CachingTorchDispatchMode. When compiling, the mode stores every op's output via _maybe_detach (line 1454 in checkpoint.py), which emits extra detach ops into the traced graph. Previously
  vanilla AC didn't use this mode so those detaches didn't exist. The extra detaches are functionally no-ops — just a side effect of unifying the AC/SAC codepaths.

File: torch/_higher_order_ops/wrap.py [hunk 5]
```diff
@@ -434,45 +418,58 @@ def __call__(self, gmod, *args, **kwargs):
...
-        with fx_traceback.preserve_node_meta():
-            from torch.utils.checkpoint import checkpoint
-
-            return checkpoint(Interpreter(gmod).run, *args, **kwargs)
     else:
-        gmod = TagActivationCheckpoint.tag_nodes(gmod, is_sac=False)
-        # Using interpreter allows preservation of metadata through torch.compile stack.
-        # TODO: We want to use the same `checkpoint(Interpreter(gmod).run, *args, **kwargs)` here
-        # as the `context_fn != None` case, but that depends on in-place op support in TorchDispatchMode + torch.compile.
-        # (for details on in-place op issue, run `test_compile_selective_checkpoint_inplace_op` unit test)
```
Comment (soulitzer) [inline torch/_higher_order_ops/wrap.py:473]: I think this comment is safe to ignore. We don't have to worry about introducing a new error because we don't actually do any inplace checks during compile anyway. We don't have to worry about silent correctness, because the tagging pass did not do any checks anyway, and plain-AC is strictly safer than SAC in terms of inplace.

File: torch/utils/checkpoint.py [hunk 6]
```diff
@@ -1305,9 +1305,10 @@ def ignore_compile_internals(cls):
         return True
 
     # Used together with _CachedTorchDispatchMode to implement SAC.
-    def __init__(self, policy_fn, storage) -> None:
+    def __init__(self, policy_fn, storage, ac_graph_id=None) -> None:
         self.policy_fn = policy_fn
         self.storage = storage
+        self.ac_graph_id = ac_graph_id
```
Comment (zou3519) [inline torch/utils/checkpoint.py:1311]: does this affect caching behavior with respect to AC?

Comment (zou3519) [reply torch/utils/checkpoint.py:1311]: to clarify, ac_graph_id goes on the nodes. Which stage of the compilation pipeline does it get onto the nodes? The two caches I'm worried about are aotautograd cache right after dynamo and fx graph cache right between aotautograd and inductor

Comment (soulitzer) [reply torch/utils/checkpoint.py:1311]: During aot autograd when the AC HOP runs. This PR doesn't change that though

Comment (soulitzer) [reply torch/utils/checkpoint.py:1311]: So this PR doesn't change what happens at the aotautograd cache right after dynamo, the fx graph cache will now see different ac_graph_id.

**Review by zou3519 — APPROVED** [review]

File: torch/_higher_order_ops/wrap.py [hunk 7]
```diff
@@ -434,45 +418,65 @@ def __call__(self, gmod, *args, **kwargs):
...
+    # instead of in AOTAutograd).
+    kwargs["preserve_rng_state"] = False
+    kwargs["context_fn"] = context_fn_with_graph_id
+    # Using interpreter allows preservation of metadata through torch.compile stack.
+    # We use a wrapper instead of passing Interpreter(gmod).run directly because
+    # checkpoint's recompute_fn captures the function in a closure. A bound method
+    # reference would keep the Interpreter alive, whose env dict retains the output
+    # tensors and prevents the autograd graph from being freed.
+    def run_with_interpreter(*args):
+        return Interpreter(gmod).run(*args)
```
Comment (soulitzer) [inline torch/_higher_order_ops/wrap.py:478]: This didn't fail previously because SAC is not applied recursively in tests (it is not supported)
